### ISSUE TITLE
Edit intervention code for new version of the model

### ISSUE THREAD

**Issue by mzimmermann-IDM** [fpsim/fpsim#352]
Once new methodtime is is implemented, we need to check the intervention code and see what we need to change to work with the new p(contraception) and time on method methodology.

**Comment by mzimmermann-IDM** [comment]
There are several types of contraceptive interventions we need to be able to implement: and intervention can change initiation of methods, switching, or discontinuation. There can be a new method or a change in method mix within existing methods. To address all of these, we will walk through these steps:

Initiation
- After women access the function for p(contraception) and get their probability, they will get a 0/1 as planned. 
- To capture initiation, we will randomly select x% (where x is specified by the user as part of the intervention definition) of women who got assigned a 0, and switch them to a 1. We will also need to add a flag to them for intervention initiators. 
- Then, when they get to method selection, the flagged women will use a new, user defined, method mix matrix. This may be a new method (in which case all women would use the new method), or some distribution of existing methods. It could even be the same as the baseline matrix if the intervention just increases use overall but not specific methods.
- We also want the option to stratify this random selection by age, parity, urban/rural, wealth quintile, education, and paid work. That way you can apply the x% differently in different subgroups as you define.

Switching
- To capture switching, after women get the 0/1 for contraceptive use, we will randomly flag y% of the 1's. Those flags will enter the new method mix matrix, which is user defined and can be 100% to a new method, similar to above. We want to be able to stratify these flags as well, with the addition of stratifying by the method they are just ending. So we can distribute the y% to women who are stopping the pill, for example.
- Getting the y% is a little more challenging here, because it depends on the baseline probability of using the 'switching from' method. So, for example, if we want to say that 5% of pill users will switch, then y = base_pill_rate*0.05. 
There is an alternative version of this, where we increase the probabilities instead of this random flagging. When Michelle and I talked through it, it seemed more complicated, but we can discuss.

Discontinuation
- For new methods, the user will have to define the time on method parameters.
- For existing methods, if an intervention decreases discontinuation, we will randomly choose women on that method (or choose them based on the stratified groups above), and increase their method time by a user defined %.

We also want to be able to implement empowerment interventions. For these, we would want to increase the number of women with a particular empowerment metric (paid work will probably be the most used example) by z%. For that year, we will select randomly z% out of women who got a 0 for that metric, and switch them to a 1. 

@MObrien-IDM did I miss anything? @pausz I am interested to hear from you how challenging you think this will be, and what other clarifications we need to make.

**Comment by mzimmermann-IDM** [comment]
@robynstuart to check if you already did this

**Comment by robynstuart** [comment]
**Initiation**: what about adding the intervention to the logistic function for calculating probabilities, e.g.
```
y = intercept + age_factors + ever_used_contra + intervention_factor
prob_use = 1 / (1+np.exp(-y))
```
that way we could get the intervention to flag people who'd been reached by the intervention, and then use the parameters in the csv files to specify what impact this would have on their probability of use.

**Switching**
One option is that when the intervention begins, we could replace the existing method switching matrix with a new one, e.g. like the attached version. In this example, 10% of people switch to the new method, and 100% of users of the new method switch to injectables when finished on the new method.

[method_mix_matrix_new_method.csv](https://github.com/user-attachments/files/16259124/method_mix_matrix_new_method.csv)


Thoughts @mzimmermann-IDM @MObrien-IDM ?

**Comment by mzimmermann-IDM** [comment]
I think I need you to talk through this one! I will be on vacation for the next two weeks, so you all can discuss without me or we can do it in August.

**Comment by mzimmermann-IDM** [comment]
@pausz @emilydriano (and @robynstuart when you get back) this is another issue we need to make sure we address before we can do analyses. Paula, I know you mentioned adding tutorials, so I think this will be part of what we need to do for that. If you can all think about what we still need to do this, that would be great!

**Comment by robynstuart** [comment]
@mzimmermann-IDM has this been done?

**Comment by mzimmermann-IDM** [comment]
Yes done!
